gips/0080-substreams-on-the-network.md

```diff
@@ -0,0 +1,112 @@
+---
+GIP: "0080"
+Title: Integrate Substreams into The Graph Network
+Authors: Alexandre Bourget <alex@streamingfast.io>
+Created: 2024-10-27
+Updated: 2024-10-27
+Stage: Draft
+Discussions-To: *To be added later*
+Category: Protocol Logic, Protocol Interfaces
+Implementations: *To be added later*
+---
+
+*This document tries to express and summarize the discussions that happened in-person at the October 2024 R&D Retreat.*
+
+# Abstract
+
+This GIP proposes integrating Substreams, a streaming-first blockchain indexing technology, into The Graph Network. This integration will enable Substreams to be fully recognized on the network and eventually receive indexing rewards.
+
+# Motivation
+
+Substreams provide significant performance improvements for indexing, particularly for large datasets and complex computations. Integrating Substreams into the network will allow developers to leverage these performance gains while maintaining the decentralization and security benefits of The Graph, as well as encouraging the market to use The Graph as its prime infrastructure provider.
+
+# Detailed Specification
+
+## Indexer Selection Algorithm (ISA)
+
+Given that there's no gateway involved with Substreams (it's direct point-to-point between consumer and provider), the existing Indexer Selection Algorithm (ISA) does not apply.  A round-robin selection of providers will be used in the front-end (https://thegraph.market) to ensure fair distribution of request loads, while still allowing consumers to choose a provider.
+
+## Payments
+
+Payments for Substreams work will be on-chain. The existing `collect` function in the staking contract will be used for payment processing.  It was deemed that no TAP was required to have Substreams collect payments and honor the promises of the protocol.
```

Comment:
TAP is not required for initial full network support, but still strongly desirable for the long term, I think it would be good to make that distinction.
IMO it would still be necessary to have some clarity on the GIP on what the MVP of payments looks like. Who calls `collect` and at what times? How is the pricing defined? How do indexer and payer agree on the price? What do indexers need to do?

Reply:
+1 one Pablo's last questions. Most readers may not understand how the staking contract works, specifically what directly calling `collect()` allows for and who can do it. The GIP lacks additional context, and references to the Permissionless Payers GIP would be appreciated, for instance. 

> How is the pricing defined? How do indexer and payer agree on the price? What do indexers need to do?
thegraph.market currently lists pricing for Substreams services, which I assume is included in that `providers.json` file. Given that .json file will be populated with the endpoints that are registered on-chain (after some health checks), some tooling is missing for Indexers to transparently advertise their rate (which seems to be tied to $ per byte). Maybe the DataEdge could be used for this as well, with a well-defined protocol, and tracked in a subgraph which eventually feeds that `providers.json`.